### PR TITLE
Staging sites: Prevent navigation on upgrades and inbox menu pages.

### PR DESCRIPTION
**Pull request by kozer**
Related to https://github.com/Automattic/dotcom-forge/issues/1881

## Proposed Changes

This pr introduces a new router middleware to prevent the navigation on all the pages under "upgrades" and "inbox" menus for staging sites.

It prevents access to all `/domains*`, `/plans*`, `/purchases*`, `/email*` and `/inbox*` sites.

## Testing Instructions

All pages mentioned below follow the following pattern:

`calypso.localhost/<pattern>/<site URL>`

## Upgrades menu

Includes the following pages:

### Plans

- `/plans`
- `/plans/compare`
- `/plans/compare/:domain`
- `/plans/features`
- `/plans/features/:domain`
- `/plans/features/:feature/:domain`
- `/plans/my-plan`
- `/plans/my-plan/trial-upgraded/:domain`
- `/plans/my-plan/:site`
- `/plans/select/:plan/:domain`
- `/plans/storage/:site`
- `/plans/upsell/:product/:site`
- `/plans/:intervalType?/:site`
- `/plans/jetpack-free/welcome`,
- `/plans/jetpack-boost/welcome`,
- `/plans/jetpack-social/welcome`,
- `/plans/complete/:site/:lang`,
- `/plans/:duration/:site`

All plan endpoints use a common `commonHandler` array, so test one, tests them all (excluded: `/plans/features/:feature/:domain` and `/plans/my-plan/trial-upgraded/:domain` where I added the middleware in there as well)

### Addons

- `/add-ons`
- `/add-ons/:site`

All add-ons endpoints use a common `commonHandler` array, so test one, tests them all.

### Domains

- All `/domains` pages 

all pages are registered via `registerStandardDomainManagementPages`, and the middleware is there, so test one of them tests them all 

### Purchases

- `/purchases`
- ` /purchases/subscriptions`
- `/purchases/:site`
- `/purchases/subscriptions/:site`
- `/purchases/subscriptions/:site/:purchaseId`
- `/purchases/subscriptions/:site/:purchaseId/cancel`
- `/purchases/subscriptions/:site/:purchaseId/confirm-cancel-domain`
- `/purchases/subscriptions/:site/:purchaseId/payment-method/add`
- `/purchases/subscriptions/:site/:purchaseId/payment-method/change/:cardId`
- `/purchases/add-payment-method/:site`
- `/purchases/billing-history/:site`
- `/purchases/billing-history/:site/:receiptId`
- `/purchases/payment-methods/:site`

All purchase endpoints use a common `commonHandler` array, so test one, tests them all.

### Email
- All `/email` endpoints
- All `/inbox` endpoints

All email endpoints use a common `commonHandler` array, so test one, tests them all.

### Email
- All `/email` endpoints
- All `/inbox` endpoints

All email endpoints use a common `commonHandler` array, so test one, tests them all.

### Test page redirection for staging sites
1. Create a staging site
2.  Try to navigate to `wordpress.com/<pattern>/<your site>`
3. Ensure you are being redirected back to the `/home` page.

### Test pages for non-staging sites
1. Create a simple site
2.  Try to navigate to `wordpress.com/<pattern>/<your site>`
3. Ensure you are  NOT being redirected back to the `/home` page.

## Pre-merge Checklist

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

### CODE REVIEW

File: client/state/selectors/test/is-staging-site.js [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+					12345: {
+						is_wpcom_staging_site: true,
+					},
+				},
+			},
+		};
+
+		expect( isStagingSite( state, 12345 ) ).toBe( true );
+	} );
+} );
```
Comment (danielbachhuber) [inline client/state/selectors/test/is-staging-site.js:41]: Thanks for introducing tests!

**Review by danielbachhuber — COMMENTED** [review]
Great work on sorting this out, @kozer 

Can you include test instructions for any non-happy paths too? Although I imagine we'll hear pretty quickly if we break something, any breakage in this area would cause a lot of pain.

**Comment by wongasy** [comment]
I've tested the email related pages and things remain intact for non-staging sites 👍

**Review by wongasy — APPROVED** [review]
I've tested the email related pages and things remain intact for non-staging sites 👍

**Comment by kozer** [comment]
> Can you include test instructions for any non-happy paths too? 
I thought about non-happy paths as well, but I couldn't find any.

To minimize the risk, I followed the exact codebase, and returned the exact same values, as `wpForTeamsP2PlusNotSupportedRedirect`.

The only way I see that this will cause a problem, is if somehow, the variable for staging would be set, without the site being a staging one, which also is something that the user can't test easily.
Another thing is if any of the above URLs need to be accessible for some reason. This would also cause a problem, and this is why I pasted all of them, for anyone who has a better overview of the URLs that we use, to take a look.

- The function I wrote, uses coalesce operation, which will not throw.

- The middlewares are a pretty standard pattern, that is already implemented by the `page` library we use.

- All the functions that I used inside the middleware, besides mine, are well tested till now (`wpForTeamsP2PlusNotSupportedRedirect` is there for a long time)

If you have found anything that could potentially cause a problem, that I'ven't thought of, please give me an example to dig further.

Thanks!

**Comment by danielbachhuber** [comment]
> If you have found anything that could potentially cause a problem, that I'ven't thought of, please give me an example to dig further.

@kozer I mostly meant to include testing instructions for non-staging sites too. However, it looks like @wongasy reviewed that for us, so this should be good to ship on Monday.

**Pull request merged**